### PR TITLE
WIP: Use stages in Travis build

### PR DESCRIPTION
**Pull request by slowli**
This PR attempts to use [Travis build stages](https://docs.travis-ci.com/user/build-stages) to streamline build process.

### CODE REVIEW

**Comment by slowli** [comment]
It seems to work now, after wasting *lots and lots* of Travis's resources 😅  The build time is now down to ~8 min (6 min if taking concurrency into account) from ~11 min. (~75% of these 8 mins are initialization and and cache management.) Note that in principle, to maximize speed, `clippy` jobs could be combined, and all 3 resulting jobs could run in parallel; it would result in ~4 min build time.

Also, there were at least 2 drawbacks in Travis config that this PR fixes:

- `cargo-audit` was installed, but not used
- `sandbox_tests/target` dir was not cached

Additionally, the PR removes `sudo: required`, [which could also speed up the build](https://docs.travis-ci.com/user/reference/trusty/#Container-based-with-sudo%3A-false):

> The advantage of running with `sudo: false` is dramatically reduced time between commit push to GitHub and the start of a job on Travis CI, which works especially well when one’s average job duration is under 3 minutes.

(As `sudo` is only needed for installing `libsodium`, it's enough to download newest `libsodium` and build it from sources. It's cached, of course, but the essence remains the same.)

**PS.** The very reason why I started the PR is because on my local PC, `cargo clippy` and `cargo test` seemed to conflict for build artifacts (i.e., executing `clippy` -> `test` -> `clippy` -> `test` would compile dependencies anew each time). This, of course, would break caching in Travis builds; and it would be completely alleviated if `clippy` and `test` jobs were split. However, I investigated a bit and found that this seems to be a problem with the previous `clippy` version (0.0.131) used by Exonum projects, which I foolishly forgot to update; the current one (0.0.142) works fine.

So, you have every right to not merge this PR; just fix the 2 issues mentioned above (and I would use `sudo: false` too, it seems nice).

File: .travis.yml [hunk 1]
```diff
@@ -1,40 +1,87 @@
...
+
+script: skip
+
+jobs:
+  include:
+  - stage: format
+    env:
+    - FEATURE=fmt
+    script:
+    - cargo audit
```
Comment (gisochre) [inline .travis.yml]: @slowli, maybe move `audit` to `lint` stage?

**Comment by slowli** [comment]
So, it seems to work, but the design needs to be finalized. I see the following options:

#### 1. Don't merge this, just implement fixes I described above.

Pros: should be fast enough.

Cons: may be a bit slower than with stages (not sure; needs to be benched to be certain). Because of lack of stages, all builds will do all checks (which may be a good thing or a bad thing; that's up to us to decide).

#### 2. Merge this, possibly with minor edits.

Pros: Stages allow to terminate the build earlier if something is wrong (e.g., formatting). Again, it may be a good thing, or a bad thing. It may be good because if linters are unsatisfied, even if tests pass, there is a risk that there are bugs introduced by bad code, which are not covered by tests.

Cons: stages may be bad, because if the submitter doesn't run tests locally, it could take multiple iterations to get to testing. Also, stages may be more resource-intensive (especially with regard to the cache size).

#### Where stages are 99% needed

In multiple-language environment, where they allow to do unit testing for each component first, and then do integration testing.

@DarkEld3r @gisochre What do you guys think?

File: .travis.yml (hunk 1 above)
Comment (slowli) [inline .travis.yml]: Sure, will do.

**Comment by gisochre** [comment]
@slowli i'd reorder then so that test stage is first everywhere. fmt and clippy are useful but more so for tidy and idiomatic code cleaned of machine-detectable obscenity.

**Comment by DarkEld3r** [comment]
@slowli I like the approach with stages, but your build time "benchmarks" isn't very clear for me. For the [last successful build](https://travis-ci.org/exonum/exonum-configuration/builds/257489605?utm_source=github_status&utm_medium=notification) Travis shows the following: 
>  Ran for 25 min 57 sec
Total time 26 min 27 sec 

Could you please trigger the build several more times (with different commits probably) in order to show build time?

**Comment by gisochre** [comment]
@slowli i meant move only `cargo audit` to `lint` stage. I like separate stages, it allows to see, what's wrong at once, without redundant details. So, for `fmt` - format stage makes sense. For `cargo audit` - audit  or lint.

**Comment by gisochre** [comment]
@slowli i meant move only `cargo audit` to `lint` stage. I like separate stages, it allows to see, what's wrong at once, without redundant details. So, for `fmt` - format stage makes sense. For `cargo audit` - audit  or lint.

**Comment by slowli** [comment]
@DarkEld3r I took [this unsuccessful build](https://travis-ci.org/exonum/exonum-configuration/builds/257512074) timing; the 25 min is caused by one of jobs building `libsodium` and `clippy` and all dependencies. The [newest build](https://travis-ci.org/exonum/exonum-configuration/builds/258162555) timed 6:40, with a particularly slow container for the final job (with a faster one, it could get ~6:00).

@gisochre In virtually all Travis configs I've encountered (with stages or not), linters precede testing, although I get your idea.

**Pull request merged**